### PR TITLE
Joey Haas

### PR DESCRIPTION
**Pull request by joeyh92989**
I was able to get the data from the endpoint and then display in the view, after first completing that in the controller i began to refactor it into facade, poro, and service structure. While i wasn't able to fully finish the refactor by getting the data into the view, i was able to complete and include tests for Poro, Service, and Facade.

### CODE REVIEW

**Comment by joeyh92989** [comment]
I have completed the refactor using facade, poro, service structure on a branch of this forked repo. I didn't merge it to this PR but was only about 10 mins away from getting the view to accept the new object ingestion of data.

File: app/controllers/food_controller.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class FoodController < ApplicationController
+  def index
+    @foods = FoodFacade.search(params[:q])
+
+    if @foods == []
+      flash[:error] = 'No food found'
+      render 'welcome/index'
```
Comment (timomitchel) [inline app/controllers/food_controller.rb:7]: Love that you have a sad path.

File: app/controllers/food_controller.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    else
+      @foods
```
Comment (timomitchel) [inline app/controllers/food_controller.rb:9]: No need to call this `@foods` explicitly here. Once you set it, it's available in the view.

File: app/facades/food_facade.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class FoodFacade
+  def self.search(search_params)
+    json = FoodService.search_foods(search_params)
+    foods =json[:foods].map { |food_info| Food.new(food_info) }
+    foods[0..9]
+  end
+  def self.total_results(search_params)
+    json = FoodService.search_foods(search_params)
+    json[:totalHits]
```
Comment (timomitchel) [inline app/facades/food_facade.rb:9]: There is a great opportunity for memoization here. Something like: 
```ruby
def self.search_foods(search_params)
json ||= FoodService.search_foods(search_params)
end
```

This way you would only ever hit the API once when you call this method and instead use the memoized response when you need to call that method again.

File: app/services/food_service.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  def self.conn
+    conn = Faraday.new(
+      url: 'https://api.nal.usda.gov/fdc/v1/foods',
+      headers: {'X-Api-Key' => ENV['fdc_api_key']}
+    )
+  end
+  def self.search_foods(search_params)
+    resp =conn.get('search', {query: search_params})
+    JSON.parse(resp.body, symbolize_names: true)
+  end
```
Comment (timomitchel) [inline app/services/food_service.rb:12]: Solid, simple service. Love where you set the API key and breaking up your arguments to the Faraday connection object onto multiple lines to improve readablity/maintainability.

File: app/poros/food.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class Food
+  attr_reader :upc,
```
Comment (timomitchel) [inline app/poros/food.rb:2]: Simple PORO. Love it.

File: coverage/.last_run.json [hunk 6]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (timomitchel) [inline coverage/.last_run.json:1]: Always a good idea to add `/coverage` to `.gitignore`.

File: spec/Facades/food_facade_spec.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  it "returns an array food with only 10 items" do
+    facade = FoodFacade.search("Sweet Potatoes")
+    expect(facade).to be_an Array
+    expect(facade.first).to be_a Food
+    expect(facade.count).to eq(10)
+  end
+  it "returns total results" do
+  facade = FoodFacade.total_results("Sweet Potatoes")
+  expect(facade).to eq(44128)
+  end
```
Comment (timomitchel) [inline spec/Facades/food_facade_spec.rb:13]: Solid testing here.

File: spec/features/welcome/index_spec.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+      expect(page).to have_content('Brand Owner: ARCHER FARMS')
+      expect(page).to have_content('Ingredients: SWEET POTATOES.')
+    end
+    it 'Sad Path it returns no results and is returned to the root' do
+      visit root_path
+      fill_in :q, with: 'asdfefgagsadfsfasdasfasfasg'
+
+      click_button 'Search'
+      expect(page).to have_content("No food found")
+    end
```
Comment (timomitchel) [inline spec/features/welcome/index_spec.rb:27]: Solid testing again. Would have loved to see you stub out the API call though.

File: spec/poros/food_spec.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+      "marketCountry": "United States",
+      "foodCategory": "Frozen Prepared Sides",
+      "allHighlightFields": "<b>Ingredients</b>: <em>SWEET</em> <em>POTATOES</em>.",
+      "score": 922.7158
+    }
+    food = Food.new(food_info)
+    expect(food.upc).to eq "492111402857"
+    expect(food.description).to eq "SWEET POTATOES"
+    expect(food.brand_owner).to eq "ARCHER FARMS"
+    expect(food.ingredients).to eq "SWEET POTATOES."
```
Comment (timomitchel) [inline spec/poros/food_spec.rb:22]: Solid testing.

File: spec/services/food_service_spec.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+require 'rails_helper'
+
+RSpec.describe 'Movie Service' do
+  it 'creates a connection object'do
+    results = FoodService.conn
+    expect(results.nil?).to eq(false)
```
Comment (timomitchel) [inline spec/services/food_service_spec.rb:6]: Also a good idea to think about what it looks like when results is `{}` an empty hash. Another op for sad path testing.

File: spec/services/food_service_spec.rb [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+  end
+  it 'returns search results'do
+  results = FoodService.search_foods("Sweet Potatoes")
+
+    expect(results[:totalHits]).to eq(44128)
```
Comment (timomitchel) [inline spec/services/food_service_spec.rb:11]: In our service we should only test the structure of how the JSON is returned, not the individual values, which are likely to change when the API changes.

**Review by timomitchel — COMMENTED** [review]
Overall, solid performance on this mid mod. LMK if you have questions about the feedback.

**Pull request closed without merging**